### PR TITLE
Get started basics

### PR DESCRIPTION
**Pull request by dvsrepo**

### CODE REVIEW

File: docs/docs/documentation/basics.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+2. Using YAML configuration files with the `Pipeline.from_yaml()`. The YAML config file in our example would be following:
+
+```yaml
+name: my-first-classifier
+head:
+    type: TextClassification
+    labels: ["positive", "negative"]
+```
+In this example, we only define the `name` and the task we want to train our model on, using the ``head`` parameter, the rest is configured from defaults. In biome.text we try to provide we sensible defaults so you don't have to configure everything just to start experimenting, but there are many things you can tune and configure.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:64]:
```suggestion
In this example, we only define the `name` and the task we want to train our model on, using the ``head`` parameter, the rest is configured from defaults. In biome.text we try to provide sensible defaults so you don't have to configure everything just to start experimenting, but there are many things you can tune and configure.
```

File: docs/docs/documentation/basics.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+1. ``lang``: the main language of the text to be tokenized (default is English). Here you can use available [spaCy model codes](https://spacy.io/usage/models/).
+2. ``segment_sentences``: enable sentence splitting for text within your input features, which is especially relevant for long text classification problems.
+3. ``text_cleaning``: simple python functions to pre-process text before tokenization. You can define your own but biome.texts provides pre-defined functions for things like cleaning up html tags or remove extra blank spaces.
+
+### Features
+Features are a central concept of the library. Building on the flexibility of AllenNLP, biome.text gives you the ability of combining [Word](../api/biome/text/features.md#wordfeatures) and [Character](../api/biome/text/features.md#charfeatures) features easily. There are many things which can be configured here: size of the embeddings, encoder type (e.g., CNNs or RNNs) for character encoding, pre-trained word vectors, and other things.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:76]:
```suggestion
Features are a central concept of the library. Building on the flexibility of AllenNLP, biome.text gives you the ability of combining [Word](../api/biome/text/features.md#wordfeatures) and [Character](../api/biome/text/features.md#charfeatures) features easily. There are many things which can be configured here: the size of the embeddings, encoder type (e.g., CNNs or RNNs) for character encoding, pre-trained word vectors, and other things.
```

File: docs/docs/documentation/basics.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+To learn more about Features configuration options and usage, see the [FeaturesConfiguration API docs](../api/biome/text/configuration.md#featuresconfiguration).
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:78]:
```suggestion
To learn more about how to configure and use Features, see the [FeaturesConfiguration API docs](../api/biome/text/configuration.md#featuresconfiguration).
```

**Comment by dvsrepo** [comment]
gracias jaja

File: docs/docs/documentation/basics.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+For defining encoders, biome.text builds on top of the `Seq2SeqEncoder` abstraction from AllenNLP, which brings many configuration possibilities, that go from RNNs to the official PyTorch Transformer implementation.
+
+### Head
+The other key component for supporting flexible transfer learning are task heads. A head defines the NLP task (e.g., text classification, token-level classification, language modelling) and specific features related to the task, for example the labels of the a text classifier (``positive`` and ``negative`` in our example).
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:88]:
```suggestion
Task heads are the other key component to support flexible transfer learning. A head defines the NLP task (e.g., text classification, token-level classification, language modelling) and specific features related to the task, for example the labels of a text classifier (``positive`` and ``negative`` in our example).
```

File: docs/docs/documentation/basics.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+| text        | label           |
+| ------------- |:-------------:|
+| I thought this was a wonderful way to spend time on a too hot summer weekend, sitting in the air con...     | positive |
+| Phil the Alien is one of those quirky films where the humour is based around the oddness of ...     | negative      |
+| ... | positive      |
+
+Data sources can also be created from [YAML configuration files](../api/biome/text/data/datasource.md#from-yaml), which might be handy for automating training and evaluation pipelines.
+
+Fields in data sources are intimately related to what the pipeline expects as input and output features. In our example, we are defining a text classification model which expects a ``text`` and a ``label``. In cases where users don't have the option to align the fields of the data with the features of the model, the class ``DataSource`` provides a `mapping`` parameter. Imagine our data set looked like this:
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:127]:
```suggestion
Columns in data sources are intimately related to what the pipeline expects as input and output features. In our example, we are defining a text classification model which expects a ``text`` and a ``label`` column. In cases where users don't have the option to align the columns of the data with the features of the model, the ``DataSource`` class provides a `mapping` parameter. Imagine our data set looked like this:
```

File: docs/docs/documentation/basics.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+Using the ``mapping`` functionality we could not only work with this data set schema by setting a mapping:
+
+```python{3}
+train_ds = DataSource(
+    source='training_data.csv',
+    mapping={'text': 'review'}
+)
+```
+
+but we could also combine both *title* and *review* to feed them as input features like so:
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:144]:
```suggestion
but we could also combine both *title* and *review* to feed them as input features:
```

File: docs/docs/documentation/basics.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+```python{3}
+train_ds = DataSource(
+    source='training_data.csv',
+    mapping={'text': ['title', 'review']}
+)
+```
+biome.text was created with semi-structured data problems in mind, so it provides specialized models for learning from structured records such as the [RecordClassification](../api/biome/text/modules/heads/classification/record_classification.md#recordclassification) head, which lets you define mappings to arbitrary input fields and combine their vector representations in a hierarchical way (e.g., combining encoders at field and record level)
+
+You can find more info about data sources and mappings in the [API documents](../api/biome/text/datasource.md#datasource).
+## Vocabulary
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:155]:
```suggestion
You can find more info about data sources and mappings in the [API documents](../api/biome/text/datasource.md#datasource).

## Vocabulary
```

File: docs/docs/documentation/basics.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+For doing NLP with neural networks, your NLP pipeline needs to turn words, subwords and/or characters into numbers. A typical process consists of tokenizing the text, and mapping word (or sub-word) tokens and maybe characters into integers or indexes. This process is often referred to as "indexing".
+
+In order to support this you need a ``Vocabulary``, which holds a mapping from tokens and characters to numerical ids (if you are interested this post provides a good [overview](https://blog.floydhub.com/tokenization-nlp/. This mapping can be learned and built during pre-training phases as is the case for word piece models and generally sub-word models like those provided by [Huggingface tokenizers](https://github.com/huggingface/tokenizers)).
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:158]:
```suggestion
In order to support this you need a ``Vocabulary``, which holds a mapping from tokens and characters to numerical ids (if you are interested, this post provides a good [overview](https://blog.floydhub.com/tokenization-nlp/)). This mapping can be learned and built during pre-training phases as is the case for word piece models and generally sub-word models like those provided by [Huggingface tokenizers](https://github.com/huggingface/tokenizers)).
```

File: docs/docs/documentation/basics.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+pipeline.create_vocabulary(VocabularyConfiguration(sources=[train_ds]))
+
+training_results = pipeline.train(
+    output="path_to_store_training_run_output",
+    training=train_ds
+)
+```
+Here we create our vocabulary from scratch using the training data source. AllenNLP provides neat abstractions for dealing with multi-feature vocabularies (e.g., chars, words, etc.) and biome.text builds on top of those abstractions to make it easy to create, reuse and extend vocabularies.
+
+To learn more about `Vocabulary` configuration options and usage, see the [VocabularyConfiguration API docs](../api/biome/text/configuration.md#vocabularyconfiguration)
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:183]:
```suggestion
To learn more about how to configure and use the `Vocabulary`, see the [VocabularyConfiguration API docs](../api/biome/text/configuration.md#vocabularyconfiguration)
```

File: docs/docs/documentation/basics.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+## Train
+Once we have everything ready, we can use Pipeline for training our model using the [train](../api/biome/text/pipeline.md#pipeline). Going back to our example:
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:186]:
```suggestion
Once we have everything ready, we can use the Pipeline to train our model using the [Pipeline.train()](../api/biome/text/pipeline.md#pipeline) method. Going back to our example:
```

File: docs/docs/documentation/basics.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+pipeline.create_vocabulary(VocabularyConfiguration(sources=[train_ds]))
+
+training_results = pipeline.train(
+    output="path_to_store_training_run_output",
+    training=train_ds
+)
+```
+Here the training output will be saved in a folder. It will contain the trained model weights and the metrics, as well as the vocabulary and a log folder for visualizing the training metrics with [tensorboard](https://www.tensorflow.org/tensorboard/).
+
+The most relevant file in this folder will be the ``model.tar.gz`` file, which bundles everything we need for loading the trained model for exploration, serving or fine-tuning.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:206]:
```suggestion
The most relevant file in this folder will be the ``model.tar.gz`` file, which bundles everything we need to explore, serve or fine-tune the trained model.
```

File: docs/docs/documentation/basics.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+```
+After loading the pipeline, you can use it for predictions:
+
+```python
+pipeline.predict(text='Good movie indeed!')
+```
+
+### Predict
+
+The [predict](../api/biome/text/pipeline.md#predict) method for our pipeline will be aligned to match our model input features (`text` in this case). Models such as the `RecordClassifier` can be used to define more fine-grained input features, such as for example an email classifier with two fields `subject` and `body`:
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:227]:
```suggestion
The arguments for the [Pipeline.predict()](../api/biome/text/pipeline.md#predict) method will be aligned to match our model input features (`text` in this case). Models such as the `RecordClassifier` can be used to define more fine-grained input features, such as, for example, an email classifier with two fields `subject` and `body`:
```

File: docs/docs/documentation/basics.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+```python
+pipeline.predict(subject='Hi!', body='Hi, hope you are well..')
+```
+
+### Explain
+Another thing we can do with a trained pipeline is to use the [explain](../api/biome/text/pipeline.md#explain) method to get the attribution of each token by means of [integrated gradients](https://arxiv.org/abs/1703.01365) for those heads that support it:
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:233]:
```suggestion
Another thing we can do with a trained pipeline is to use the [Pipeline.explain()](../api/biome/text/pipeline.md#explain) method to get the attribution of each token by means of [integrated gradients](https://arxiv.org/abs/1703.01365) for those heads that support it:
```

File: docs/docs/documentation/basics.md [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+```python
+pipeline.explain(text='Good movie indeed!')
+```
+
+### Serve
+Pipeline use the awesome [FastAPI](https://fastapi.tiangolo.com/) library to give you simple REST endpoint with the [serve](../api/biome/text/pipeline.md#serve) method, which provides methods aligned with your input features (e.g., a method accepting `subject` and `body` parameters):
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:239]:
```suggestion
Pipeline uses the awesome [FastAPI](https://fastapi.tiangolo.com/) library to give you a simple REST endpoint with the [Pipeline.serve()](../api/biome/text/pipeline.md#serve) method, which provides methods aligned with your input features (e.g., a method accepting `subject` and `body` parameters):
```

File: docs/docs/documentation/basics.md [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+
+```python
+pipeline.serve(port=9090)
+```
+
+### Explore
+In order to support users with fine-grained error analysis and empower them to improve their models with informed decisions, the [explore](../api/biome/text/pipeline.md#explore) method launches a UI (inside your notebook if you are using Jupyter).
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:246]:
```suggestion
In order to support users with fine-grained error analysis and empower them to improve their models with informed decisions, the [Pipeline.explore()](../api/biome/text/pipeline.md#explore) method launches an UI (inside your notebook if you are using Jupyter).
```

File: docs/docs/documentation/basics.md [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+)
+```
+
+This search-based UI can help:
+
+ - to quickly identify difficult examples in the training, validation, test or any other data set.
+ - to inform users about corner cases and which labels need more examples.
+ - to find the values for confidence thresholds to use the model in production.
+ - to help domain experts understand the strengths and limitations of the model.
+ - in summary, get a sense of the model interacting with data.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:261]:
```suggestion
 - in summary, get a sense of how the model interacts with your data.
```

File: docs/docs/documentation/basics.md [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+
+:::tip
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:263]:
```suggestion
:::tip Tip
```

This was a comment by Leire to override the default "TIP" header (all capital) to "Tip"

File: docs/docs/documentation/basics.md [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+For the UI to work you need a running [Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/install-elasticsearch.html) instance. We recommend installing [Elasticsearch with docker](https://www.elastic.co/guide/en/elasticsearch/reference/7.7/docker.html#docker-cli-run-dev-mode).
+:::
+
+### Training and transfer learning
+
+After loading a pipeline, you can keep on training it for the same task with new data:
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:269]:
```suggestion
After loading a pipeline, you can continue the training for the same task with new data:
```

File: docs/docs/documentation/basics.md [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+```python
+training_results = pipeline.train(
+    output="path_to_store_further_training",
+    training=new_training_ds
+)
+```
+:::tip
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:276]:
```suggestion
:::tip Tip
```
see comment above

File: docs/docs/documentation/basics.md [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+    TextClassification,
+    labels = ["horror_movie", "comedy", "drama", "social"]
+)
+
+training_results = pipeline.train(
+    output="path_to_fine_tuned",
+    training=categories_training_ds
+)
+```
+Here we just use the [set_head](../api/biome/text/pipeline.md#set-head) method to set a new tasks head which classifies film categories instead of review sentiment.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:294]:
```suggestion
Here we just use the [Pipeline.set_head()](../api/biome/text/pipeline.md#set-head) method to set a new task head which classifies film categories instead of review sentiment.
```

File: docs/docs/documentation/basics.md [hunk 21]
```diff
@@ -0,0 +1,303 @@
...
+
+The more usual pre-training a language model + fine-tuning on downstream tasks is also supported by using a [LanguageModelling](../api/biome/text/modules/heads/language_modelling.md#languagemodelling) head for pre-training.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:296]:
```suggestion
The more common "pre-training a language model + fine-tuning on downstream tasks" is also supported by using a [LanguageModelling](../api/biome/text/modules/heads/language_modelling.md#languagemodelling) head for pre-training.
```

File: docs/docs/documentation/basics.md [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+
+
+## Next steps
+
+The best place to start is the tutorials sections.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:301]:
```suggestion
The best place to start is the tutorials section.
```

File: docs/docs/documentation/basics.md [hunk 23]
```diff
@@ -0,0 +1,303 @@
...
+
+Also you find detailed [API docs](../api/) and use the search feature to find out more about specific concepts and features.
```
Comment (dcfidalgo) [inline docs/docs/documentation/basics.md:303]:
```suggestion
You can also find detailed [API docs](../api/) and use the search bar on top to find out more about specific concepts and features.
```

**Comment by dvsrepo** [comment]
> Very very very nice!! I enjoyed reading it a lot, just a few inline comments from my side. feel free to implement them or to leave them where they belong ...

I've included all your suggestions, good review thanks!
